streams/src/main/java/org/apache/kafka/streams/state/internals/metrics/RocksDBMetricsRecorder.java

```diff
@@ -17,117 +17,231 @@
-        memtableMinFlushTimeSensor = RocksDBMetrics.memtableMinFlushTimeSensor(streamsMetrics, metricContext);
-        memtableMaxFlushTimeSensor = RocksDBMetrics.memtableMaxFlushTimeSensor(streamsMetrics, metricContext);
         writeStallDurationSensor = RocksDBMetrics.writeStallDurationSensor(streamsMetrics, metricContext);
         blockCacheDataHitRatioSensor = RocksDBMetrics.blockCacheDataHitRatioSensor(streamsMetrics, metricContext);
         blockCacheIndexHitRatioSensor = RocksDBMetrics.blockCacheIndexHitRatioSensor(streamsMetrics, metricContext);
         blockCacheFilterHitRatioSensor = RocksDBMetrics.blockCacheFilterHitRatioSensor(streamsMetrics, metricContext);
-        bytesReadDuringCompactionSensor = RocksDBMetrics.bytesReadDuringCompactionSensor(streamsMetrics, metricContext);
         bytesWrittenDuringCompactionSensor =
             RocksDBMetrics.bytesWrittenDuringCompactionSensor(streamsMetrics, metricContext);
-        compactionTimeAvgSensor = RocksDBMetrics.compactionTimeAvgSensor(streamsMetrics, metricContext);
-        compactionTimeMinSensor = RocksDBMetrics.compactionTimeMinSensor(streamsMetrics, metricContext);
-        compactionTimeMaxSensor = RocksDBMetrics.compactionTimeMaxSensor(streamsMetrics, metricContext);
+        bytesReadDuringCompactionSensor = RocksDBMetrics.bytesReadDuringCompactionSensor(streamsMetrics, metricContext);
         numberOfOpenFilesSensor = RocksDBMetrics.numberOfOpenFilesSensor(streamsMetrics, metricContext);
         numberOfFileErrorsSensor = RocksDBMetrics.numberOfFileErrorsSensor(streamsMetrics, metricContext);
     }
 
-    public void addStatistics(final String storeName,
-                              final Statistics statistics,
-                              final StreamsMetricsImpl streamsMetrics,
-                              final TaskId taskId) {
-        if (statisticsToRecord.isEmpty()) {
-            init(streamsMetrics, taskId);
-        }
-        if (statisticsToRecord.containsKey(storeName)) {
-            throw new IllegalStateException("A statistics for store " + storeName + "has been already registered. "
-                + "This is a bug in Kafka Streams.");
-        }
-        statistics.setStatsLevel(StatsLevel.EXCEPT_DETAILED_TIMERS);
-        statisticsToRecord.put(storeName, statistics);
-    }
-
     public void removeStatistics(final String storeName) {
         final Statistics removedStatistics = statisticsToRecord.remove(storeName);
         if (removedStatistics != null) {
             removedStatistics.close();
+            log.debug("Removed statistics for store {}", storeName);
+        } else {
+            throw new IllegalStateException("No statistics for store \"" + storeName
+                + "\" found. This is a bug in Kafka Streams.");
+        }
+        if (state == State.RUNNING && statisticsToRecord.isEmpty()) {
+            state = State.NOT_RUNNING;
+            thread.interrupt();
+            waitForThreadToDie();
```

Comment:
One caveat is that when we are closing the Kafka Streams instance with a specified timeout value, this function may violate that timeout and wait for longer time since we call `thread.join()` without a timeout value.

Reply:
Just to clarify, that add/removeStatistics do not need to be thread-safe since they would only be called within a single thread at a given time right?